### ISSUE TITLE
General feedback on the Nav 2 API usability project

### ISSUE THREAD

**Issue by InMatrix** [flutter/uxr#6]
If you have any feedback on the scope of the [Nav 2 API usability project](https://github.com/flutter/uxr/wiki/Navigator-2.0-API-Usability-Research) and the approach we've taken, please leave your comments below. Thanks!

**Comment by InMatrix** [comment]
I'd like to clarify the intent and scope of the project, prompted by some questions in #4. 

First, the primary goal of this project is to design or _endorse_ an easy-to-use package for implementing common navigation patterns through usability research. While @johnpryan has prototyped the [page_router](https://github.com/johnpryan/page_router) package, we're not attached to a first-party solution. At the moment, it's mostly used as a tool for investigating where the sweet spot might be between API usability and flexibility. We are aware of several community efforts around simplifying Navigator 2.0, and a potential outcome of this project is to recommend one of them in a way similar to how [the provider package](https://flutter.dev/docs/development/data-and-backend/state-mgmt/simple) became the recommended state management solution for relatively simple Flutter apps.

But we need to establish a standard to make sure whatever we recommend or choose to build ourselves meets the needs of Flutter users and strikes the right balance between usability and flexibility. This project is about establishing that standard and developing a method for measuring whether a package has reached that standard. The first step, which is discussed in this issue, is to define the set of navigation scenarios most Flutter developers would need and design the high-level API for these scenarios. The biggest question we're looking into right now is whether nested routing should be supported by a high-level navigation API or not, and how much supporting that could negatively impact the usability of the API in other scenarios such as deeplinking. 

The scenario definition work drives the [next few activities](https://github.com/flutter/uxr/wiki/Navigator-2.0-API-Usability-Research#planned-work) in our research and design process. We'd like to include a community package in our API usability studies alongside `page_router`, so we can understand how users react to different API design approaches. The results of those studies will be published in the wiki section of this repo. If you're an author of a navigation package, we'd love to get in touch and learn about your future plans. 

Last, we recognize that usability was insufficiently considered in the design process of Navigator 2.0 and we'd like to develop a process for not only facilitating the simplification of Navigator 2.0 through high-level APIs but also aid future API design projects in Flutter.

**Comment by slovnicki** [comment]
Thank you for opening this discussion. I'm the creator of [beamer](https://pub.dev/packages/beamer) and would like to contribute to this investigation.

Beamer was created during my development of a fairly large mobile/web app, trying to abstract the concepts from [this article](https://medium.com/flutter/learning-flutters-new-navigation-and-routing-system-7c9068155ade) by John Ryan and an [official design doc](https://docs.google.com/document/d/1Q0jx0l4-xymph9O6zLaOY4d_f7YFpNWX_eGbzYxr9wY/edit).

I ended up forming an idea of `BeamLocation` which represents the `RouterDelegate.currentConfiguration`. It holds everything related to `path` and is responsible for providing `pages`  to the `Navigator`. The `pages` list is dynamically created, conforming to the rules the developer has set having examined the available `path` properties. Common usage is that the developer groups "related" `Page`s into a custom implementation of `BeamLocation` to which the user can then `beamTo` and the appropriate stack will be built. User can also update it's current stack via `updateCurrentLocation`.

The end result is having all the rules for app's routing in a single place and developer needs not to worry much about "navigation state". It ends up being similar to named routes, but now we have "named locations". There are also some additional features such as `beamBack` and guards.

The package is still in early stages and with lots of improvements planned, all of which can be monitored at [issue board](https://github.com/slovnicki/beamer/issues). I have made examples of some scenarios and they are available as gifs in README and as full code samples at [example](https://github.com/slovnicki/beamer/blob/master/example/lib/main.dart) (the original books example) and [examples](https://github.com/slovnicki/beamer/tree/master/examples).

Any contribution is welcome and I believe this discussion will be of great help for future work.

**Comment by nguyenxndaidev** [comment]
Hi,

First of all, thanks flutter team so much for listening to developers and making the next big change to Navigator API.

I am an Angular developer with some years of experience. Even I still have some issues with angular router in my projects, **I love angular router a lot**. With the experience when working with angular router, I checked all the navigation packages I found on pub.dev, but to be honest, **I don't like all of them**. I will definitely not use them for my projects. Therefore, last month, I started to create my own project to copy angular router design to flutter. I do not publish it yet, and will not publish it because now I believe flutter team will soon deliver a better official API.

So, I would like to give my feedback from what I learned when I tried to copy angular router design to flutter.

From my point of view, flutter could copy almost stuffs from the angular router with improvements for usability and flexibility. Below are some improvements I am trying to do while copying angular router to flutter:
* By default, use [Ant Path pattern](http://ant.apache.org/manual/dirtasks.html#patterns) for the balance of usability and flexibility in routing configuration. I think Ant Path is the easiest way to declare a route, I personally don't like to use a complicated solution like RegExp, but it would be nice if the API still allow developers to change to the way they want for matching routes flexibly.
* Instead of [Route Guard](https://angular.io/guide/router#preventing-unauthorized-access), I would like to have a more general solution, for example, **Route Interceptor**. Each route should accepts multiple route interceptors, running in a chain with configured ordering. Child-routes (or sub-routes) should inherits all route interceptors from parent routes. The interceptor could be used for checking if user is allowed to route from route A to route B. If not, developers can choose what to do, ex.
  * showing a dialog to tell user buying PRO version
  * redirect user to route C
  * stop or continue with next route interceptor

  Not only that, interceptor is a good place to create/destroy required instances for the route or integrate with an injection solution. For example, I could use [GetIt](https://github.com/fluttercommunity/get_it) to create/destroy an [injection scope](https://github.com/fluttercommunity/get_it#scopes) for any specific route and its sub-routes.
* Up and Back action
  * Up action: for navigating to upper level in the the route hierarchy. This should be the default behavior of the Up button in the [AppBar](https://api.flutter.dev/flutter/material/AppBar-class.html).
  * Back action: for navigating to previous route in history or close the popup dialog. This should be the default behavior of Back button.
  * Up and Back action, should be callable anywhere when developers need it, without the need of user interaction on Up and Back buttons.
  * Allow developer to change the default behavior of Up and Back buttons mentioned above globally with 1 or 2 lines of code. But still allow to overwrite the default behavior for specific pages when needed.
* Sub-routes configuration should be inherited from parent routes and can be overwritten. For example, these configuration could be inherited:
  * Route Interceptors
  * Default behavior of Up and Back button
  * Transition animation
  * Etc.
* Support working with modularization projects
  I am working in a Banking project, which is very big project. Therefore, we must split our resources into multiple teams.
  For example, team A maintains the main app and some main modules. We have other teams B, C, D, each team maintains some modules in separated git repositories.
  We want each module control its own routing. Routing inside the module is totally independent and considered it is working from root route. When it joins the main app, it belongs to a sub-route of a parent route in the main app.
  For example, module A has routes: `/`, `/products`, `/appoinments`. When we integrate with main app, the main app will configure to have `/customers/:id` as parent route of module A. So the final routes in our app becomes: `/customers/:id`, `/customers/:id/products`, `/customers/:id/appoinments` when we integrate module A to the main app.
* Work nicely with stream
  Some teams in my project want to control navigation with state. So instead of using the provided methods (ex. push(), pop() or whatever), they want to have a bloc to control the app current route state by themselves. Therefore, they want to use `StreamBuilder` widget to provide new route state and with this, the app navigate to next route automatically when route state is changed. The state should contain the current route, the history stack (not the route hierarchy stack in the current implementation of Navigator 2.0).
* Transition animation
  * Up and Back action should be configurable to use different transition animations, so users know they are going up to the route hierarchy or back to the previous route in the history.
  * Transition should not only be configurable globally, but also for specific route and sub-routes.

**Comment by idkq** [comment]
Hi, many thanks for this, excited to make this better. Few general comments:

- If 'Nav 2.0' would have been named 'Nav Advanced' instead, I believe there would be no 'complaints'. Nav 2 made developers believe that Nav 1 would/could be obsolete at some point or it was inferior. So people, especially people building new apps they logically jumped straight to 2.0. So, I guess there was a 'branding' issue here. If it was named 'Nav Advanced' only people looking for missing features would perhaps investigate it.
- Be careful with definitions of Route/Page/Screen. I don't think they are clearly defined in Nav 2 and easily understood. For example, look at the docs for [class Route<T>](https://api.flutter.dev/flutter/widgets/Route-class.html). Definition: "An abstraction for an entry managed by a Navigator." This is not a good definition "An abstraction for an entry". That said, clearly define what is a route, how it is different from other concepts, how it plays out, etc. Also [class Page<T>](https://api.flutter.dev/flutter/widgets/Page-class.html) "Describes the configuration of a Route.". It is not a definition, but it says what it does.

Regarding standards, in my opinion:

- Should be **intuitive** - developers could (although maybe shouldn't) be able to code without reading much docs. Instead, using code completion, etc. Why? Because all developers have experience with navigation, it is not something new for anyone.
- Should be **pre-assembled** - offering off-the-shelve templates or scenarios for common use cases
- Should be **flexible** - allowing to create custom patterns of navigation.
- Should be **high-level** - requiring minimum line of code, 1 line is perfect, 2 is great, 3 is good, 4 is okay, 5 is hmm

Hope it helps as a start.

**Comment by lulupointu** [comment]
Hi,

I am the developer of the [vrouter](https://pub.dev/packages/vrouter) plugin. My approach was based on Vue.js router.

The basic idea is to make a routing tree at the root of the application, describing your paths and the widgets which belong to each paths. I believe this is working very well in terms of simplicity and flexibility.

Here is a basic example:
```
VRouter(
  routes: [
    VStacked(
      path: '/login',
      widget: LoginWidget(),
    ),
    VStacked(
      path: '/profile',
      widget: ProfileWidget(),
      subroutes: [
        VStacked(
          path: 'settings', // This matches the path '/profile/settings'
          widget: ProfileWidget(),
        ),
      ],
    ),
    VRouteRedirector(
      path: ':_(.*)', // This matches any path
      redirectTo: '/login',
    )
  ],
)
```

Has you can see, this makes for something very condensed. However there are a lot of other functionalities such as:
* Custom global or local transitions
* Nesting route (either by stacking them or nesting them in the Fluttery sense)
* React to route changes
* Custom Handling pop event
* Name 
* Aliases
* ...

I created a website to explain how to use it, in hope to make it easy for anyone to use: [vrouter.dev](https://vrouter.dev)

**Comment by agordeev** [comment]
Many of us use `flutter_bloc` for state management. It'd be great if the docs include an example of how to properly use Navigator 2.0 with cubits/blocs.

**Comment by lulupointu** [comment]
I created the 6 scenarios describes in [Navigation scenario storyboards](https://github.com/flutter/uxr/files/5953028/PUBLIC.Flutter.Navigator.API.Scenarios.-.Storyboards.pdf) using [vrouter](https://pub.dev/packages/vrouter).

I tried to produce minimal yet realistic implementation of the described situations.
Please tell me if you want me to comment the code more than I did, or if I misunderstood any of the situations.

<details>
 <summary>1.Deep Linking - Path Parameters</summary>

```dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

main() {
  runApp(
    VRouter(
      routes: [
        VStacked(
          path: '/home',
          widget: HomePage(),
          subroutes: [
            VStacked(
              path: r'/book/:id(\d+)', // We match any id number
              widget: DetailPage(),
            ),
          ],
        ),
        VRouteRedirector(path: ':_(.*)', redirectTo: '/home'),
      ],
    ),
  );
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Center(child: Text('Path Parameters')),
      ),
      body: ListView.builder(
        itemBuilder: (BuildContext context, int index) {
          final bookColor = Color((Random().nextDouble() * 0xFFFFFF).toInt()).withOpacity(1.0);
          return Container(
            color: bookColor,
            height: 50,
            child: TextButton(
              onPressed: () => VRouterData.of(context)
                  .push('/book/$index', routerState: '${bookColor.value}'),
              child: Text('Book number $index'),
            ),
          );
        },
      ),
    );
  }
}

class DetailPage extends StatefulWidget {
  @override
  _DetailPageState createState() => _DetailPageState();
}

class _DetailPageState extends State<DetailPage> {
  Color bookColor;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Path Parameters'),
      ),
      body: VNavigationGuard(
        afterEnter: (_, __, ___) => setState(() {
          bookColor = Color(int.tryParse(VRouterData.of(context).historyState));
        }),
        child: Container(
          color: bookColor,
          child: Center(
              child:
              Text('This is book number ${VRouteData.of(context).pathParameters['id']}')),
        ),
      ),
    );
  }
}
```

</details>

<details>
 <summary>2.Deep Linking - Query Parameters</summary>

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

main() {
  runApp(
    VRouter(
      routes: [
        VStacked(
          path: '/products',
          widget: ProductsPage(),
        ),
        VRouteRedirector(path: ':_(.*)', redirectTo: '/products'),
      ],
    ),
  );
}

class ProductsPage extends StatefulWidget {
  final itemCount = 200;

  @override
  _ProductsPageState createState() => _ProductsPageState();
}

class _ProductsPageState extends State<ProductsPage> {
  bool sortAlphabetically = true;

  @override
  Widget build(BuildContext context) {
    return VNavigationGuard(
      // After we enter or update, we check sortAlphabetically queryParameter to see if the
      // UI is synced with the url
      afterEnter: (_, __, ___) => setState(() => sortAlphabetically =
          VRouteData.of(context).queryParameters['sortAlphabetically'] == 'true'),
      afterUpdate: (_, __, ___) => setState(() => sortAlphabetically =
          VRouteData.of(context).queryParameters['sortAlphabetically'] == 'true'),
      child: Scaffold(
        appBar: AppBar(
          title: Center(child: Text('Query Parameters')),
        ),
        body: Column(
          children: [
            TextButton(
              onPressed: () => VRouterData.of(context).pushReplacement('/products',
                  queryParameters: {'sortAlphabetically': (!sortAlphabetically).toString()}),
              child: Row(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Text('Click to sort'),
                  Icon(sortAlphabetically ? Icons.arrow_downward : Icons.arrow_upward),
                ],
              ),
            ),
            Expanded(
              child: ListView.builder(
                itemCount: widget.itemCount,
                itemBuilder: (context, index) {
                  return Container(
                    alignment: Alignment.center,
                    decoration: BoxDecoration(
                      border: Border(top: BorderSide(color: Colors.black)),
                    ),
                    height: 50,
                    child: Text(
                        'Book number ${sortAlphabetically ? index : (widget.itemCount - index - 1)}'),
                  );
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

<details>
 <summary>3. “Dynamic Linking”</summary>

```dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

class GameInfo {
  final String username;
  final int id;

  GameInfo({@required this.username, @required this.id});
}

// This global variable is set yet we enter a game room
GameInfo gameInfo;

main() {
  runApp(
    VRouter(
      routes: [
        VStacked(
          path: '/game',
          widget: GamePage(),
          subroutes: [
            VStacked(
              path: r':id(\d+)', // Matches any digit
              name: 'gameRoom',
              widget: GameRoomPage(),
              // Before we enter a gameRoom, we check that the gameInfo is set
              // if not we redirect to "/game"
              beforeEnter: (vRedirector) async => (gameInfo == null)
                  ? vRedirector.push('/game',
                      routerState: vRedirector.newVRouteData.pathParameters['id'])
                  : null,
            )
          ],
        ),
        // Redirect any miss-typed url to "/game"
        VRouteRedirector(path: ':_(.*)', redirectTo: '/game'),
      ],
    ),
  );
}

class GamePage extends StatefulWidget {
  @override
  _GamePageState createState() => _GamePageState();
}

class _GamePageState extends State<GamePage> {
  // The username is whatever the user types in the TextFild
  String username;

  // The game id is gotten from the history state if any
  // This happens when the user goes back to the previous page
  int gameId;

  @override
  Widget build(BuildContext context) {
    return VNavigationGuard(
      // in afterEnter we try to get the gameId from the historyState
      afterEnter: (_, __, ___) => (VRouterData.of(context).historyState != null)
          ? setState(() => gameId = int.parse(VRouterData.of(context).historyState))
          : null,
      child: Material(
        child: Column(
          children: [
            TextField(
              decoration: InputDecoration(
                border: OutlineInputBorder(borderSide: BorderSide(color: Colors.teal)),
              ),
              onChanged: (username) => this.username = username,
              onSubmitted: (_) =>
                  (gameId == null) ? createGameRoom() : joinGameRoom(id: gameId),
            ),
            TextButton(
              onPressed: () => (gameId == null) ? createGameRoom() : joinGameRoom(id: gameId),
              child: Text((gameId == null) ? 'Create game room' : 'Join game room $gameId'),
            ),
            if (gameId == null && gameInfo != null)
              TextButton(
                onPressed: () => joinGameRoom(id: gameInfo.id),
                child: Text('Go back game room ${gameInfo.id}'),
              ),
          ],
        ),
      ),
    );
  }

  // This is used when the user want to create a fresh new room
  void createGameRoom() {
    gameInfo = GameInfo(username: username, id: Random().nextInt(1000));
    VRouterData.of(context).pushNamed('gameRoom', pathParameters: {'id': '${gameInfo.id}'});
  }

  // This is used when the user want to join a game room which he has the game id
  void joinGameRoom({@required int id}) {
    gameInfo = GameInfo(username: username, id: id);
    VRouterData.of(context).pushNamed('gameRoom', pathParameters: {'id': '${gameInfo.id}'});
  }
}

class GameRoomPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Material(
        child: Text('Welcome ${gameInfo.username}, you are in the room number ${gameInfo.id}'
            '\nCopy the url to invite other to join!'),
      ),
    );
  }
}
```

</details>

<details>
 <summary>4. “Skipping Stacks”</summary>

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

main() {
  runApp(
    VRouter(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        textButtonTheme: TextButtonThemeData(
          style: ButtonStyle(
            padding: MaterialStateProperty.all(EdgeInsets.all(20.0)),
            backgroundColor: MaterialStateColor.resolveWith((states) => Colors.black),
            textStyle: MaterialStateProperty.all(TextStyle(fontSize: 16, color: Colors.white)),
          ),
        ),
      ),
      routes: [
        VStacked(
          path: '/search',
          widget: SearchResultPage(),
          subroutes: [
            VStacked(
              path: ':productName',
              name: 'productPage',
              widget: ProductPage(),
            ),
          ],
        ),
        VStacked(
          // In parenthesis is a regExp which allows us to limit the productType
          //    - This is less verbose than creating multiple route
          //    - We can access the product type with the path parameter productType
          path: r'/:productType(book|shoes|pillow)',
          widget: CategoryPage(),
          subroutes: [
            VStacked(
              path: ':productName',
              name: 'productPage',
              widget: ProductPage(),
            ),
          ],
        ),
        VRouteRedirector(path: ':_(.*)', redirectTo: '/search'),
      ],
    ),
  );
}

// We do not implement a search bar but only the page on which a user would be
// after he/she typed in a certain url
class SearchResultPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Center(child: Text('Dynamic linking')),
      ),
      body: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 30.0, horizontal: 20),
            child: Align(
              alignment: Alignment.topLeft,
              child: Text(
                'Here is the result for you search "stars":',
                style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
              ),
            ),
          ),
          ProductList(
            products: allProducts.where((product) => product.name.contains("stars")).toList(),
          )
        ],
      ),
    );
  }
}

class ProductPage extends StatefulWidget {
  @override
  _ProductPageState createState() => _ProductPageState();
}

class _ProductPageState extends State<ProductPage> {
  Product product;

  @override
  Widget build(BuildContext context) {
    product ??= allProducts.firstWhere(
      (product) => product.name == VRouteData.of(context).pathParameters['productName'],
    );

    return Scaffold(
      appBar: AppBar(
        title: Center(child: Text('Dynamic linking')),
        actions: [
          IconButton(
            icon: Icon(Icons.search),
            onPressed: () => VRouterData.of(context).push('/search'),
          )
        ],
      ),
      body: Hero(
        tag: product.name,
        child: Material(
          child: Container(
            color: productTypeColor(product.type),
            child: Center(
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Text('Type: ${productTypeToString(product.type)}'),
                  Text(product.name),
                  SizedBox(
                    height: 20,
                  ),
                  TextButton(
                    onPressed: () =>
                        VRouterData.of(context).push('/${productTypeToString(product.type)}'),
                    child: Text(
                      'See all ${productTypeToString(product.type)}',
                      style: TextStyle(color: Colors.white),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

class CategoryPage extends StatefulWidget {
  @override
  _CategoryPageState createState() => _CategoryPageState();
}

class _CategoryPageState extends State<CategoryPage> {
  List<Product> products = [];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Center(child: Text('Dynamic linking')),
        actions: [
          IconButton(
            icon: Icon(Icons.search),
            onPressed: () => VRouterData.of(context).push('/search'),
          )
        ],
      ),
      body: VNavigationGuard(
        afterEnter: (_, __, ___) => getProducts(),
        afterUpdate: (_, __, ___) => getProducts(),
        child: ProductList(
          products: products,
        ),
      ),
    );
  }

  void getProducts() {
    setState(() {
      products = allProducts
          .where((product) =>
              productTypeToString(product.type) ==
              VRouteData.of(context).pathParameters['productType'])
          .toList();
    });
  }
}

class ProductList extends StatelessWidget {
  final List<Product> products;

  const ProductList({Key key, @required this.products}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Wrap(
        spacing: 20,
        runSpacing: 20,
        alignment: WrapAlignment.center,
        children: products,
      ),
    );
  }
}

class Product extends StatelessWidget {
  final String name;
  final ProductTypes type;

  const Product({Key key, @required this.name, @required this.type}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: () => VRouterData.of(context).push(name),
      child: Hero(
        tag: name,
        child: Material(
          child: Container(
            height: 200,
            width: 200,
            color: productTypeColor(type),
            child: Center(
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Text('Type: ${productTypeToString(type)}'),
                  Text(name),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

enum ProductTypes { book, shoe, pillow }

Color productTypeColor(ProductTypes itemType) {
  switch (itemType) {
    case ProductTypes.book:
      return Colors.greenAccent;
    case ProductTypes.shoe:
      return Colors.redAccent;
    case ProductTypes.pillow:
      return Colors.blueAccent;
    default:
      return Colors.black;
  }
}

String productTypeToString(ProductTypes itemType) {
  switch (itemType) {
    case ProductTypes.book:
      return 'book';
    case ProductTypes.shoe:
      return 'shoes';
    case ProductTypes.pillow:
      return 'pillow';
    default:
      return 'unknown';
  }
}

final bookProducts = <Product>[
  Product(name: 'Book stars', type: ProductTypes.book),
  Product(name: 'old book', type: ProductTypes.book),
  Product(name: 'eBook', type: ProductTypes.book),
];

final shoesProducts = <Product>[
  Product(name: 'Shoes of stars', type: ProductTypes.shoe),
  Product(name: 'Only One Shoe', type: ProductTypes.shoe),
  Product(name: 'Shoe fleur', type: ProductTypes.shoe),
];

final pillowProducts = <Product>[
  Product(name: 'Best pillow', type: ProductTypes.pillow),
  Product(name: 'Pillow stars', type: ProductTypes.pillow),
  Product(name: 'This is not a pillow', type: ProductTypes.pillow),
];

final allProducts = <Product>[...bookProducts, ...shoesProducts, ...pillowProducts];
```

</details>

<details>
 <summary>5. Login/Logout/Sign-up Routing</summary>

```dart
// Note that this example uses shared_preferences to persistently store the fact that one is logged in

import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:vrouter/vrouter.dart';

main() {
  runApp(
    VRouter(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        textButtonTheme: TextButtonThemeData(
          style: ButtonStyle(
            padding: MaterialStateProperty.all(EdgeInsets.all(20.0)),
            backgroundColor: MaterialStateColor.resolveWith((states) => Colors.black),
            textStyle: MaterialStateProperty.all(TextStyle(fontSize: 16, color: Colors.white)),
          ),
        ),
      ),
      // beforeEnter is called to check that we do not access something we are not supposed to
      beforeEnter: (vRedirector) async {
        if (vRedirector.to != '/login') {
          final sharedPreferences = await SharedPreferences.getInstance();
          // If isLoggedIn is false, we are not logged in, so we get redirected
          if (!(sharedPreferences.getBool('isLoggedIn') ?? false)) vRedirector.push('/login');
        }
      },
      routes: [
        VStacked(
          path: '/login',
          widget: LoginPage(),
          // If we are already connected, no need to go here so we redirect to "/home"
          beforeEnter: (vRedirector) async {
            final sharedPreferences = await SharedPreferences.getInstance();
            if (sharedPreferences.getBool('isLoggedIn') ?? false) vRedirector.push('/home');
          },
        ),
        VStacked(path: '/home', widget: HomePage()),
        VRouteRedirector(path: ':_(.*)', redirectTo: '/home')
      ],
    ),
  );
}

class LoginPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: TextButton(
        onPressed: () async {
          final sharedPreferences = await SharedPreferences.getInstance();
          await sharedPreferences.setBool('isLoggedIn', true);
          VRouterData.of(context).push('/home');
        },
        child: Text('Click to connect!'),
      ),
    );
  }
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Material(
      child: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('Well done, you are connected!'),
            SizedBox(height: 20),
            TextButton(
              onPressed: () async {
                final sharedPreferences = await SharedPreferences.getInstance();
                await sharedPreferences.setBool('isLoggedIn', false);
                VRouterData.of(context).push('/login');
              },
              child: Text('Click to logout!'),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

<details>
 <summary>6. Nested Routing</summary>

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

main() {
  runApp(
    VRouter(
      routes: [
        VStacked(
          key: ValueKey('MyScaffold'),
          widget: MyScaffold(),
          subroutes: [
            VChild(
              path: '/${HomePage.label}',
              name: HomePage.label,
              widget: HomePage(),
            ),
            VChild(
              path: '/${SchedulePage.label}',
              name: SchedulePage.label,
              widget: SchedulePage(),
              subroutes: [
                VChild(
                  path: '${MainViewTab.label}',
                  name: '${MainViewTab.label}',
                  widget: MainViewTab(), // This will never be used in practice
                ),
                VChild(
                  path: '${KanbanViewTab.label}',
                  name: '${KanbanViewTab.label}',
                  widget: KanbanViewTab(), // This will never be used in practice
                ),
              ],
              // "/schedule" should never be accessed, if it is we redirect to "/schedule/main"
              beforeEnter: (vRedirector) async =>
                  vRedirector.pushReplacement('/${SchedulePage.label}/${MainViewTab.label}'),
            ),
            VChild(
              path: '/${BudgetPage.label}',
              name: BudgetPage.label,
              widget: BudgetPage(),
            ),
            VChild(
              path: '/${TeamPage.label}',
              name: TeamPage.label,
              widget: TeamPage(),
            ),
          ],
        ),
        VRouteRedirector(path: ':_(.*)', redirectTo: '/${HomePage.label}'),
      ],
    ),
  );
}

class MyScaffold extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Material(
      child: Row(
        children: [
          LeftNavigationBar(),
          Expanded(child: VRouteElementData.of(context).vChild),
        ],
      ),
    );
  }
}

class LeftNavigationBar extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.max,
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        LeftNavigationBarTextButton(label: HomePage.label),
        LeftNavigationBarTextButton(label: SchedulePage.label),
        LeftNavigationBarTextButton(label: BudgetPage.label),
        LeftNavigationBarTextButton(label: TeamPage.label),
      ],
    );
  }
}

class LeftNavigationBarTextButton extends StatelessWidget {
  final String label;

  const LeftNavigationBarTextButton({Key key, @required this.label}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final isSelected = (VRouteElementData.of(context).vChildName == label);
    return TextButton(
      onPressed: () {
        if (!isSelected) VRouterData.of(context).pushNamed(label);
      },
      style: ButtonStyle(
        alignment: Alignment.topLeft,
        padding: MaterialStateProperty.all(EdgeInsets.all(20.0)),
      ),
      child: Text(
        label,
        style: TextStyle(
          fontWeight: FontWeight.bold,
          fontSize: 16,
          color: isSelected ? Colors.blueAccent : Colors.black,
        ),
      ),
    );
  }
}

class HomePage extends StatelessWidget {
  static final label = 'home';

  @override
  Widget build(BuildContext context) {
    return Center(child: Text('This is the $label page.'));
  }
}

class SchedulePage extends StatefulWidget {
  static final label = 'schedule';

  @override
  _SchedulePageState createState() => _SchedulePageState();
}

class _SchedulePageState extends State<SchedulePage> with SingleTickerProviderStateMixin {
  TabController tabController;

  @override
  void initState() {
    tabController = TabController(length: 2, vsync: this);
    super.initState();
  }

  @override
  void dispose() {
    tabController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    // We check the url to see if the current selected tab is the same as the one
    // in the url
    final tabBarViewIndex =
        VRouteElementData.of(context).vChildName == MainViewTab.label ? 0 : 1;
    if (tabBarViewIndex != tabController.index) tabController.animateTo(tabBarViewIndex);

    // We listen to the tabController and push a new route if the index changes
    tabController.addListener(() {
      if (tabBarViewIndex != tabController.index)
        VRouterData.of(context).pushReplacementNamed(
            (tabController.index == 0) ? MainViewTab.label : KanbanViewTab.label);
    });

    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        Center(child: Text('This is the ${SchedulePage.label} page.')),
        Container(
          color: Colors.greenAccent,
          child: TabBar(
            controller: tabController,
            labelStyle: TextStyle(fontWeight: FontWeight.bold, fontSize: 20),
            labelPadding: const EdgeInsets.all(20.0),
            tabs: [
              Center(
                child: Text(MainViewTab.label),
              ),
              Center(
                child: Text(KanbanViewTab.label),
              ),
            ],
          ),
        ),
        Expanded(
          child: TabBarView(
            controller: tabController,
            children: [
              MainViewTab(),
              KanbanViewTab(),
            ],
          ),
        )
      ],
    );
  }
}

class MainViewTab extends StatelessWidget {
  static final label = 'main';

  @override
  Widget build(BuildContext context) {
    return Center(child: Text('This is the $label tab view.'));
  }
}

class KanbanViewTab extends StatelessWidget {
  static final label = 'kanban';

  @override
  Widget build(BuildContext context) {
    return Center(child: Text('This is the $label tab view.'));
  }
}

class BudgetPage extends StatelessWidget {
  static final label = 'budget';

  @override
  Widget build(BuildContext context) {
    return Center(child: Text('This is the $label page.'));
  }
}

class TeamPage extends StatefulWidget {
  static final label = 'team';

  @override
  _TeamPageState createState() => _TeamPageState();
}

class _TeamPageState extends State<TeamPage> {
  final List<String> names = [
    'Liam',
    'Olivia',
    'Noah',
    'Emma',
    'Oliver',
    'Ava',
    'William',
    'Sophia',
    'Elijah',
    'Isabella',
    'James',
    'Charlotte',
    'Benjamin',
    'Amelia',
    'Lucas',
    'Mia',
    'Mason',
    'Harper',
    'Ethan',
    'Evelyn'
  ];

  String nameEditBeingShown;

  @override
  Widget build(BuildContext context) {
    final nameBeingEdited = VRouteData.of(context).queryParameters['edit'];
    if (nameBeingEdited != nameEditBeingShown) {
      if (nameBeingEdited == null) {
        Navigator.of(context, rootNavigator: true).pop();
      } else {
        if (nameEditBeingShown != null) {
          Navigator.of(context, rootNavigator: true).pop();
        }
        WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
          showDialog(
            barrierDismissible: false,
            context: context,
            builder: (_) => Material(
              color: Colors.transparent,
              child: InkWell(
                onTap: () => VRouterData.of(context).push('/team'),
                child: Center(
                  child: Container(
                    padding: EdgeInsets.all(20.0),
                    color: Colors.white,
                    child: Text('You are editing $nameBeingEdited'),
                  ),
                ),
              ),
            ),
          );
        });
      }
      nameEditBeingShown = nameBeingEdited;
    }

    return Center(
      child: ConstrainedBox(
        constraints: BoxConstraints(maxWidth: 200),
        child: Column(
          children: [
            Text('This is the ${TeamPage.label} page.'),
            Expanded(
              child: ListView(
                children: names
                    .map(
                      (name) => ListTile(
                        title: Text(name),
                        trailing: IconButton(
                          icon: Icon(Icons.edit),
                          onPressed: () {
                            VRouterData.of(context)
                                .push('/team', queryParameters: {'edit': name});
                          },
                        ),
                      ),
                    )
                    .toList(),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

If you are tired of copy/pasting, you can also clone [this repo](https://github.com/lulupointu/vrouter_navigator_scenarios)

I hope to hear want you think !

**Comment by InMatrix** [comment]
@lulupointu That's awesome. Thank you so much. I played with vrouter yesterday and I found it really easy to get started. The way it handles nested routes feels easy to understand as well. The scenario implementations you shared will greatly help us conduct the comparative analysis we plan on doing next (#7).

**Comment by jacobaraujo7** [comment]
Hello!
flutter_modular use Navigator 2.0 as well. 
https://pub.dev/packages/flutter_modular/versions/3.0.0-nullsafety.45

**Comment by nguyenxndaidev** [comment]
Another idea is to use the way of Spring Framework MVC to handle routes with annotation. In the controller method, developer can decide to build a widget or redirect to another route.

**Comment by esDotDev** [comment]
The one thing that seems missing from the proposal, and honestly you see this quite often in the Widget library itself, is lack of research into prior art.

imo we should be looking hard at what came in React and Vue, and simple either borrowing those hardened proven paradigms, or seeing where you can additively improve them. Flutter team has a habit of completely re-inventing the wheel, instead of realizing that many smart people have already tackled these problems, and just borrowing from that wisdom.

**Comment by esDotDev** [comment]
Looking at the latest implementation of  React Router (v4) 2 things stand out immediately:
1. It is self-contained, not bundled inside an App components, using 3 delegates to do everything. This means it is very simple to grok.
2. Within 1m, you have the urlBar changing, and there is literally no way to mess it up.

In terms of dev experience, this is vastly more focused, and easier to understand:
https://reactrouter.com/

If flutter can get anywhere close to this ease of use, I'm sure developers will be extremely pleased.

**Comment by nguyenxndaidev** [comment]
I think the approach of react router is the most similar to the way of flutter widget. It worth to check it.

**Comment by TahaTesser** [comment]
@InMatrix 
We just triaged a proposal https://github.com/flutter/flutter/issues/77360, we like to know if it possible or could be implemented in the future
Thanks!

**Comment by slovnicki** [comment]
> @InMatrix
> We just triaged a proposal [flutter/flutter#77360](https://github.com/flutter/flutter/issues/77360), we like to know if it possible or could be implemented in the future
> Thanks!

It's definitely possible and in fact quite natural in Navigator 2.0, as the developer has control over Navigator's entire stack of pages.

**Comment by InMatrix** [comment]
@TahaTesser This is a good example of the Skipping Stacks (#5) use case. We probably should consider incorporating that in our scenario deck. Cc: @jackkim9

**Comment by HosseinYousefi** [comment]
Hey everyone!

I'm the author of [Yeet](https://pub.dev/packages/yeet). I'm focusing on 1. Keeping things extremely simple for first time developers (something that just works!) and 2. Treating web as a first class citizen.

I'm currently teaching a [3 week course on flutter](https://github.com/HosseinYousefi/2021-CS208) and spending 2 days to really teach something like navigator 2.0 is not a viable option for beginners. 

My solution was to create something that feels like yet another tree (we're using flutter after all). I use a simple depth first search to find the first match and populate the page stack using that. It's still quite experimental but the main functionalities do work.

**Comment by rrousselGit** [comment]
There are use-cases that I think are not covered by the current examples:

- Windows resizing / responsive design.
  Web/Desktop application, due to the platform relying on windows instead of full-screen pages, may want to handle resizing the window to use a different layout.
  The difficulty is in preserving the navigation and application state during layout switch.

- Next-route pre-loading
  For performance, applications may want to lazy-load data or widgets (such as with the new [deferred components](https://flutter.dev/docs/perf/deferred-components))
  The side-effect is that when navigating to a new route, these objects may not be loaded yet, which lead to showing progress indicators.
  To improve the user experience, a common practice is trying to preload the "next" route.
  An example would be:
   - an application has allows reading a book
   - when the user is on page x, load in background page x+1.
     Page pre-loading includes deferred components, images, network requests, ...
   - page x may want to disable or delay navigation to x+1 while x+1 is loading.

**Comment by InMatrix** [comment]
@rrousselGit These are great considerations. Some of them came up in earlier discussions, especially with the regard to the interplay between routing and state management. We decided to keep it simple in this evaluation, but routing API designers should take the issues you raised into consideration. I'd love to see responses to your questions from package authors.

**Comment by esDotDev** [comment]
Of these 4 issues, 1 of them is outside the hands of package authors, which is preloading an off-stage widget or routes. Flutter needs better support in general for keeping things alive, or pre-cached when off stage. 

The other 3 issues all point to why state-first navigation is a really bad idea. When you have 2 sources of truth (url path OR state), it gets very messy very fast. Because web urls are a fact of life, we can not get away from them, which means you either choose URL-first approach, or you have to deal with the huge headache that is reconciling state and url. 

It becomes clear very quickly in production, that state needs to go, and URL-first is the only sane path forward.

It reminds me of 10yrs ago, before unidirectional app state caught on. It was never exactly clear who was changing the state. Eventually everyone got on page that this was a big headache and mess, and agreed that state-changes should be top down, with the views binding to the models but not changing them directly.

I think it would be a good idea at some point, for the Flutter team to take a step back, understand that state-first nav is actually not a good approach for the majority of apps, and maybe engineer a URL-first approach that is more robust and flexible than Nav 1.

**Comment by rrousselGit** [comment]
> Of these 4 issues, 1 of them is outside the hands of package authors, which is preloading an off-stage widget or routes. Flutter needs better support in general for keeping things alive, or pre-cached when off stage.

I disagree

For example with state-first approach, we could potentially support:

```dart
SomeNavigator(
  routes: [
     if (page != null) ...[
       MyRoute(path: '/book/$selectedBookId/page/$page'),
       Offstage(child: MyRoute(path: '/book/$selectedBookId/page/${page + 1}')),
     ],
  ]
)
```

**Comment by rrousselGit** [comment]
I would also not call url-first as "imperative"

For me the imperative APIs in Navigator 1 are not "pushNamed" functions but rather:
- onGenerateRoute, because it is called only once. So it is not possible to update a route once it was built
- `push(context, Route)`, for similar reasons. Once pushed, the route cannot be updated.

This is in opposition to things like `MaterialApp.routes`/`MaterialApp.home`, where you can rebuild the `MaterialApp` widget with different routes/home

**Comment by esDotDev** [comment]
> > Of these 4 issues, 1 of them is outside the hands of package authors, which is preloading an off-stage widget or routes. Flutter needs better support in general for keeping things alive, or pre-cached when off stage.
> 
> I disagree
> 
> For example with state-first approach, we could potentially support:
> 
> ```dart
> SomeNavigator(
>   routes: [
>      if (page != null) ...[
>        MyRoute(path: '/book/$selectedBookId/page/$page'),
>        Offstage(child: MyRoute(path: '/book/$selectedBookId/page/${page + 1}')),
>      ],
>   ]
> )
> ```

Sure, but this is not very scalable and will quickly turn into a mess. It's much simpler and more flexible to just give MyRoute someway it can cache the next route, and then show that cached route when it needs. But flutter doesn't support this as everything needs to be 'on-stage' before it can be initialized.

Though, to be fair, probably more effective anyways to preload the data that the view needs, and then when the view loads in, there is nothing to fetch.

**Comment by esDotDev** [comment]
> For me the imperative APIs in Navigator 1 are not "pushNamed" functions but rather:

I think you're getting far away from the classic definition of imperative here. I just mean that routes are explicitly navigated to via a function (like `pushNamed`), rather than implicitly shown based on whatever the current state of the application is, which does not scale well at all, and provides no solution for routes that are not tied to state, you end up with cruft like `appModel.isSettingsPanelOpen` and inevitably clash as you add more and more of these flags.

**Comment by idkq** [comment]
>> The other 3 issues all point to why state-first navigation is a really bad idea.

I agree but I would say that much more cons to state navigation exist than reconciling state and URL. By state navigation I mean having several variables controlling which screen should be rendered and be presented as the active screen. The variables are inter-related and inter-dependent. For example if isLoggedIn is false present login screen, if searchingProduct is not null present Search screen. But if isLoggedIn is true and searchingProduct is not null, present Home Screen. 

Is this easy to maintain? No.

State navigation is preposterous in 90% of the cases. State navigation is a conditional stacked navigation, that is, the navigation and screen is rebuild based on a series of 'if' conditions. 

How many variables do you need to determine which screen is active? Only one. Not one per screen. One can argue that the existing state variables used for other purposes are being re utilized for navigation but, no, that is not a benefit.

Let's compare it to an useful state utilization. Say you have a Switch that filters some results. The state of the switch can modify many aspects of that screen including the color of the switch itself. The switch state is binary but the widget tree can change significantly based on the state.

Going back to navigation the process is similar but useless. The change of navigation state most of the times occur when an event is triggered (press button etc). Like the switch, the screen is either presented or not (other screen in the stack would be presented). The problem here is that absolutely nothing depends on the state of the navigation in 90% of the cases. The entire screen is assumed to be active for any widgets to be shown. The navigation state is never consumed for any other means. It is a waste of time for over engineering it.

So in summary, state navigation uses many more variables than necessary, makes the code hard to maintain, makes state available but for no widget to consume, and requires syncing with URL.

**Comment by johnpryan** [comment]
The original [snippets](https://github.com/flutter/samples/tree/master/navigation_and_routing/old_snippets) written for [Learning Flutter's new navigation and routing system](https://medium.com/flutter/learning-flutters-new-navigation-and-routing-system-7c9068155ade) used a state-driven approach, which is probably not the best-practice for routing, as is being discussed.

There's a new sample in the flutter/samples repo, [navigation_and_routing](https://github.com/flutter/samples/tree/master/navigation_and_routing), which demonstrates a general-purpose way to use the current URL path as the app state for navigation.

**Comment by idkq** [comment]
@johnpryan Excellent job! I really like this way better. Loved it this complete example and implementation overall. Kudos!

Here is what I don't like (no criticism to the code whatsoever, but conceptually, in line with the previous discussion on variables and state manage):

On `BookstoreNavigator`, `selectedBook` and `selectedAuthor` are driving the navigation stack: it dictates which screens will be stacked - this makes no sense to me. First of all, there is no 'selected' book or author concept. Instead, the user is simply 'viewing' a book page or author page. The difference is important because if we treat every (or some) page-parameters as conceptual active/inactive object, things can get messy.

Secondly, the only consumers of `selectedBook` and `selectedAuthor` are their respective screens, similarly to a single web-page request, this should be scoped down, and not up. Why do we need to parse and use it in the navigator? We need it only to put the `MaterialPage` into the stack. So, `selectedBook` and `selectedAuthor` are not being used anywhere else in the app except their own screens. It means that as the app grows it would have several useless variables and ifs in this `BookstoreNavigator` class for no reason. If only `BookDetailsScreen` cares about param book, it should be limited to that scope and not other parts of the app.

An argument can be made that another process could change the state of that variable hence it would activate the screen. But this is not what is happening in this app and it is an unlikely use case overall. As an analogy, you are currently reading this GitHub webpage. Changing the state of a parameter of another webpage does not rebuild your navigation stack.

A better approach in my view would be the navigation stack be built based on the active `RouteStateScope` path hierarchy. Think of it as a Site Map structure. For example: 

- '/' would return 1 page: the index/main view. 
- '/books' returns 2 pages: main view + `BooksScreen`
- '/book/0' returns 3 pages: main view + `BooksScreen` + `BookDetailsScreen`

For that to happen a linkage of some sort would need to be implemented, such as 'parent' for each page. So that when pop() is used it moves to the 'parent' route.

That could also be used on the outer `Navigator` in `BookstoreScaffoldBody` to replace the ifs. In case there are no parents, show the main view.

Please lmk if I'm missing something... I just can't see a use case in which bringing these variables up to the widget tree is applicable unless rare cases of monitoring, services and sensors.

**Comment by slovnicki** [comment]
@idkq I agree with your points.
Have you seen how [beamer](https://github.com/slovnicki/beamer) works?

The scoping of navigation variables for use cases in large applications is exactly what beamer aims to solve with the concept of `BeamLocation`s. The process looks something like this:

- `RouteInformation` comes into the app
- `BeamerDelegate` (`RouterDelegate`) takes it and chooses the appropriate `BeamLocation` to which to pass this `RouteInformation`. In this process, the pure URI (e.g. `/books/2`) stays in `BeamerDelegate` (without additional variables like "selected book" from path parameter).
- `BeamLocation` creates its own `state` based on pure `RouteInformation` passed in above step. At this point, this `state` can be anything what developer defines, but there is also a default `BeamState` data object that will extract `pathParameters` and such variables. The `state` just needs to mix with `RouteInformationSerializable` in order for `BeamerDelegate` to understand what route to report when `currentBeamLocation` updates.
- `BeamLocation` builds a stack of pages to put into `Navigator.pages` based on its `state`.

After this is completed and we are at `/books/2` with a page stack of `[BooksScreen, BookDetailsScreen]`, we can navigate 

- within our `BooksBeamLocation` by updating its `state` (e.g. setting `selectedBook` to null, which would effectively make a `pop`)
- via `beamToNamed('/articles/3')` (and some other ways) which will run the above process again, choosing the `ArticlesBeamLocation` this time, for example.

Here, the information about "selected books" or "selected article" is localized within the appropriate `BeamLocation`; above just the relevant page stack, and not above everything. The same can be done for scoping certain additional `Providers` and such (e.g. `CartProvider` for all shop-related page stacks), but this is just an additional feature of `BeamLocation`s, not related to navigation.

---

As for your idea of "1 path segment = 1 page", this is the default behavior of beamer, but can also be overridden if developer needs. When creating a navigation package, some default behavior needs to be picked and this is, in my opinion also, the most sensible one.

---

**NOTE:** Some of my statements above might be confusing (especially regarding the types I mention) if compared to the current (`v0.14.1`) README of beamer. This is because I'm in the middle of implementing these improvements. Basically

- when I say `RouteInformation`, the `v0.14.1` documentation would say `BeamState`
- there is no `RouteInformationSerializable` concept in `v0.14.1`. `BeamLocation`s can still have custom states in `v0.14.1`, but this is not so clean nor clear as it will be in `v0.15.0`

**Comment by idkq** [comment]
@slovnicki I like that beamer builds the page stack purely on the URI info. 

Do you see a valid scenario in which the page stack structure changes at runtime? If not, I believe `BeamLocation` could be further simplified (perhaps a new `SimpleBeamLocation`?) that does not have the ifs such as `if (state.uri.pathSegments.contains('books'))`. Just an idea.

The **build method of the super class** would be in charge of determining which pages to put or remove (not put) in the stack.

```dart
@override
  List<BeamPage> buildPages(BuildContext context, BeamState state) => [
        BeamPage(
          key: ValueKey('home'),
          child: HomeScreen(),
          path: '/'
        ),
        BeamPage(
          key: ValueKey('books'),
          child: BooksScreen(),
          path: '/books'
        ),
        BeamPage(
          key: ValueKey('book-${state.pathParameters['bookId']}'),
          child: BookDetailsScreen(
            bookId: state.pathParameters['bookId'],
            ),
          path: '/books/:bookid'
      ];
```

The buildPages would loop trough all `BeamPage` list and remove the ones not needed and keep others based on BeamPage.path match with current URI. Maybe `key` could be used for path url if you don't want to create a new attribute.

One problem is when you have a bifurcation, such as '/' -> '/books' and '/' -> '/authors'. For that perhaps use a new `BeamLocation` or create a new class `BeamPageBifurcation` that takes a list of `BeamPage`.

It is a minor thing but just for maintenance leaving the URI resolving at the backstages as opposed to be controlled by the developer. 

Potentially even `path` attribute could be omitted. The logic would be to match the number/count of `BeamPage` to the level of the URI. For example, '/' = 1, '/xxxx' = 2, '/xxxx/yyyy' = 3 (count of forward slashes plus one if any text). Not sure for readability though.

```dart
@override
  List<BeamPage> buildPages(BuildContext context, BeamState state) => [
        BeamPage(  //resolves from '/'
          key: ValueKey('home'),
          child: HomeScreen(),
        ),
        BeamPage( //resolves from '/books'
          key: ValueKey('books'),
          child: BooksScreen(),
        ),
        BeamPage( //resolves from '/books/:bookId'
          key: ValueKey('book-${state.pathParameters['bookId']}'),
          child: BookDetailsScreen(
            bookId: state.pathParameters['bookId'],
            ),
      ];
```

**Comment by slovnicki** [comment]
@idkq When you say "SimpleBeamLocation", are you referring to **the** `SimpleBeamLocation` that is a part of beamer? This special `BeamLocation` is used by `SimpleLocationBuilder` that does exactly what you describe; determines the appropriate `BeamPage`s to put into stack by examining path segments (+ `/` as first page)

**Comment by idkq** [comment]
@slovnicki Got it, exactly.. you are ahead on the game! Just saw in your example here https://github.com/flutter/uxr/blob/master/nav2-usability/scenario_code/lib/deeplink-pathparam/deeplink_pathparam_beamer.dart

**Comment by johnpryan** [comment]
> On BookstoreNavigator, selectedBook and selectedAuthor are driving the navigation stack: it dictates which screens will be stacked - this makes no sense to me. First of all, there is no 'selected' book or author concept. Instead, the user is simply 'viewing' a book page or author page. The difference is important because if we treat every (or some) page-parameters as conceptual active/inactive object, things can get messy.

The `if (selectedBook != null)` statement could be replaced with `if (pathTemplate == '/book/:bookId')` - would that make more sense?

> Secondly, the only consumers of selectedBook and selectedAuthor are their respective screens, similarly to a single web-page request, this should be scoped down, and not up. Why do we need to parse and use it in the navigator? We need it only to put the MaterialPage into the stack. So, selectedBook and selectedAuthor are not being used anywhere else in the app except their own screens. It means that as the app grows it would have several useless variables and ifs in this BookstoreNavigator class for no reason. If only BookDetailsScreen cares about param book, it should be limited to that scope and not other parts of the app.

You could certainly make `BookDetailsScreen` and `AuthorDetailsScreen` obtain the selected book / author, instead of having BookstoreNavigator provide them.

> An argument can be made that another process could change the state of that variable hence it would activate the screen. But this is not what is happening in this app and it is an unlikely use case overall. As an analogy, you are currently reading this GitHub webpage. Changing the state of a parameter of another webpage does not rebuild your navigation stack.

I don't think this was the reasoning. The reason it works this ways is that `AuthorDetailsScreen` requires an `Author`, so someone needs to provide that. It just so happens that `BookstoreNavigator` is building the `AuthorDetailsScreen`. Of course `AuthorDetailsScreen` could obtain the selected author based on the route ID obtained by `routeState.route.parameters['authorId']` and then obtain the selected author from `LibraryScope.of(context).allAuthors`, but as it's currently written the AuthorsScreen is decoupled from the navigation state.

> A better approach in my view would be the navigation stack be built based on the active RouteStateScope path hierarchy. Think of it as a Site Map structure.

This is what packages like vrouter, beamer, and auto_route do.

> For that to happen a linkage of some sort would need to be implemented, such as 'parent' for each page. So that when pop() is used it moves to the 'parent' route.
> That could also be used on the outer Navigator in BookstoreScaffoldBody to replace the ifs. In case there are no parents, show the main view.

I'm not sure what you mean by this. Are you talking about how `onPopPage` would be implemented?

**Comment by idkq** [comment]
> The if (selectedBook != null) statement could be replaced with if (pathTemplate == '/book/:bookId') - would that make more sense?

Yes. But I rather have it done automatically by another class and getting rid of the conditions entirely. So it can be reused in any app.

Here's an example https://github.com/idkq/samples -> navigation_and_routing

Take a look at the newly created class `SimpleNavigator`. Few comments:

- Replaces the two default `Navigator`
- In a nutshell it simply filters the pages that should be stacked, based on a matching regex of the url
- Used keys to store url regex (not ideal). Ideally we would have another class holding the matching regex. 

>  The reason it works this ways is that AuthorDetailsScreen requires an Author, so someone needs to provide that.

Yeah, in the example I pasted, the `AuthorDetailsScreen` would require an `Author` in the constructor even if not being used, which let to a hack as you can see in the example.

**Comment by johnpryan** [comment]
> Take a look at the newly created class SimpleNavigator.

Most of the higher-level packages that were studied abstract away the `Page` objects too, not just the Navigator (beamer uses `BeamPage` and vrouter uses `VWidget` for example). Something to consider, since using the key needs to match the route as currently written.

> Yeah, in the example I pasted, the AuthorDetailsScreen would require an Author in the constructor even if not being used, which led to a hack as you can see in the example.

This is why its better to only build the page objects when they are needed. In your version, each page is being constructed because the `if (selectedBook)` statements were removed.

**Comment by idkq** [comment]
@johnpryan It was a quick (and honestly dirty) example I posted, just to demonstrate. Here is a better updated version. https://github.com/idkq/samples - no keys dependency, instead abstracting Page.

All logic is automatic and behind the scenes. No need for the ifs and stacking manually. So the user code is cleaner and easier to maintain as the app grows.

`StackingLogic` can be subclassed and customized to handle complex cases. In this demo, there aren't any L2+ multi level stacks but in the real world almost certainly there are. Here is what I mean by multi level stack:

Here is the entire app Tree Map:

There are two Navigators: Outer / Inner Navigators

- Signin Screen
  - Outer: /signin
  - Inner: null
-  BooksList Screen
   - Outer: scaffold 
   - Inner: /books
- AuthorsList Screen
  - Outer: scaffold
  - Inner: /authors
- Settings Screen 
  - Outer: scaffold
  - Inner: /settings
- BooksDetail Screen
  - Outer: scaffold + /book/:bookId
  - Inner: /books
- AuthorDetail Screen
  - Outer: scaffold + /author/:authorId
  - Inner: /authors

The most we have is two stacked screens in one nav on BooksDetail and AuthorDetail. Still they are not really two levels because scaffold is always present by default. No logic is needed to decide to show/not show a scaffold. So for this demo, stacking is straightforward.

A 'truly two levels' would be (three levels in total), for example:

- AuthorProfile Screen (new)
  - Outer: scaffold + /author/:authorId + /author/:authorId/profile

Here '/author/:authorId/profile' must be stacked after '/author/:authorId'. When popped  '/author/:authorId/profile' must return to '/author/:authorId'. To automate I see two ways/approaches (open to suggestions on other approaches):

- **Using the path as hierarchy**: both urls have '/author/:authorId' as common ancestors. So the stack could be built automatically using the url only with `AncestorStackingLogic`
- **Using named parent:** each page can point to a parent in which when the child is stacked, the parent must also be stacked. This is also true recursively for n number of parents, until the root. This is done by `ParentStackingLogic`.

`SimpleNavigator` requires a `StackingLogic` class in which it delegates the stacking algo.

SIDE NOTE: when using push & link, things are not working properly on the original code. I'm not sure if that use intentionally to show that push does not have the Route info or if it is a bug. Reproduce: 

1. In Books -> Click 'Left Hand...' book0
2. Click View Author Link
3. Click 'Left Hand...' book0 again
4. Click View Author Link again

Link does not work. This happen because in `AuthorDetailsScreen` you uses push. When mixing push and link the app does not behave correctly.

**Comment by MateusAmin** [comment]
Maybe out of place, but how would one use snack bars with the navigation?

A common scenario is you confirm your email after signup by following a link back the app. At that point I notify the person that their email is verified and redirect them to the authenticated app page.

Notifying with a snack bar seems most appropriate. You don't need them to take and action and want to get the started ASAP.

But doing it in nav 2.0 does not seem straight forward. Single data point; I haven't figured it in a few hours spent.

**Comment by oodavid** [comment]
Regarding pre-loading the next screen, are we mostly interested in pre-rendering that screen, or making sure the data has been loaded?

Rendering should be pretty snappy, so if it's data, then my issue / question about inherited data for child routes might have some overlap:

https://github.com/flutter/uxr/issues/72
